scylla/src/policies/load_balancing/default.rs

```diff
@@ -602,6 +588,77 @@ impl DefaultPolicy {
         routing_info
     }
 
+    /// Checks for misconfiguration and warns if any is discovered.
+    fn pick_sanity_checks(&self, routing_info: &ProcessedRoutingInfo, cluster: &ClusterState) {
+        if let Some(preferred_dc) = self.preferences.datacenter() {
+            // Preferred DC + no datacenter failover + SimpleStrategy is an anti-pattern.
+            if let Some(ref token_with_strategy) = routing_info.token_with_strategy {
+                if !self.permit_dc_failover
+                    && matches!(
+                        token_with_strategy.strategy,
+                        Strategy::SimpleStrategy { .. }
+                    )
+                {
+                    warn!("\
+Combining SimpleStrategy with preferred_datacenter set to Some and disabled datacenter failover may lead to empty query plans for some tokens.\
+It is better to give up using one of them: either operate in a keyspace with NetworkTopologyStrategy, which explicitly states\
+how many replicas there are in each datacenter (you probably want at least 1 to avoid empty plans while preferring that datacenter), \
+or refrain from preferring datacenters (which may ban all other datacenters, if datacenter failover happens to be not possible)."
+                    );
+                }
+            }
+
+            // Verify that the preferred datacenter is actually present in the cluster.
+            // If not, shout a warning.
+            if !cluster.all_nodes.iter().any(|node| {
+                node.datacenter
+                    .as_deref()
+                    .is_some_and(|dc| dc == preferred_dc)
+            }) {
```

Comment:
This iterates over potentially all nodes (even in a happy case). I don't think it is a good idea performance-wise to have such checks in the hot path.

Reply:
Can't you use ReplicaLocator for that? I see it has `ReplicationInfo` which has:
```rust
    /// We keep a separate token ring for each datacenter.
    /// Each datacenter's ring contains only nodes from this datacenter.
    /// It makes it simpler and more efficient to calculate the NetworkTopologyStrategy replicas.
    datacenters: HashMap<String, DatacenterNodes>,
```

Reply:
Good catch!